### PR TITLE
WIP: Add WidgetStream

### PR DESCRIPTION
**Pull request by martinRenou**
First working example of a "WidgetStream", the API can be changed if you don't like it. I didn't see how to make it a MediaStream object, that's why I created a separate widget. It works fine with basic widgets like `Slider`, with map widgets from `ipyleaflet` and it works fine with `pythreejs`, but I couldn't make it work with the example I had with `ipyvolume`.

### CODE REVIEW

**Comment by martinRenou** [comment]
It's currently not possible to record it or to stream it through the chat, there is still work to be done.

**Comment by martinRenou** [comment]
@maartenbreddels I implemented the WidgetStream the way we discussed and it works fine :) It works great with pythreejs (because I use the captureStream method on the webgl canvas) and other MediaStream as input. It works fine (1fps) with widgets which are not canvas nor videos (for example slider or ipyleaflet map), I didn't try for ipyvolume.

File: js/src/webrtc.js [hunk 1]
```diff
@@ -56,6 +57,115 @@ var MediaStreamView = widgets.DOMWidgetView.extend({
...
+                            resolve(this.canvas.mozCaptureStream());
+                        }
+                    } else {
+                        reject(new Error('captureStream not supported for this browser'));
+                    }
+                });
+            };
+
+            return new Promise((resolve, reject) => {
+                updateStream(view.el);
```
Comment (martinRenou) [inline js/src/webrtc.js]: We need a method for stopping the infinite recursive call, maybe a `stream` attribute set to true by default, and if it's false we stop the stream with html2canvas.

Comment (maartenbreddels) [reply js/src/webrtc.js]: I think we should stop once the widget is closed.

Comment (martinRenou) [reply js/src/webrtc.js]: When is the widget considered closed? This code is part of the javascript model, so isn't it "alive" as long as the python object exists?

Comment (martinRenou) [reply js/src/webrtc.js]: Ok thanks :)

**Comment by maartenbreddels** [comment]
Please rebase again master.
Also, we may want to put in a fps trait, for the max fps it will do capturing using setTimeout or requestAnimationFrame.

**Comment by martinRenou** [comment]
Yes I thought about putting a fps trait, I'll do it :)

**Comment by martinRenou** [comment]
Actually this fps trait should be defined for all MediaStream classes

**Comment by martinRenou** [comment]
I named it `max_fps`, I thought it was a better name because it is not possible to reach say 25fps with html2canvas

File: ipywebrtc/webrtc.py [hunk 2]
```diff
@@ -39,12 +39,31 @@ class MediaStream(DOMWidget):
...
+
+@register
+class WidgetStream(MediaStream):
+    """Represents a widget media source.
+    """
+    _model_name = Unicode('WidgetStreamModel').tag(sync=True)
+    _view_name = Unicode('WidgetStreamView').tag(sync=True)
+
+    widget = Instance(DOMWidget, allow_none=False).tag(sync=True, **widget_serialization)
+    max_fps = Int(60).tag(sync=True)
```
Comment (maartenbreddels) [inline ipywebrtc/webrtc.py]: There is also a difference between passing an argument and not passing an argument. We could have allow_none=True, and have None as default value.

Comment (martinRenou) [reply ipywebrtc/webrtc.py]: As you wish, I can do that if you want :) Do you agree with the fact that we should put a fps attribute for each MediaStream class ?

Comment (martinRenou) [reply ipywebrtc/webrtc.py]: Maybe in an other PR

**Comment by maartenbreddels** [comment]
Congrats! Great work 👍

**Comment by martinRenou** [comment]
Thanks :)

**Pull request merged**